Clarify control flow in step generator

swgillespie opened the pull request:
This commit introduces a new type, 'controlFlow', which indicates to
callers whether or not it should continue the operation that it was
doing before it called a particular function. The variants of
'controlFlow', 'Bail' and 'Continue', indicate to the caller what it
should do. Components can signal 'Bail' when a diagnostic is logged and
the plan should no longer continue.

swgillespie commented:
This PR is a small part of https://github.com/pulumi/pulumi/issues/1694 that tries to introduces some coding conventions that, if we like them, can bring to the rest of the engine. In order to realize the full benefit of this we'll probably need to do a lot more plumbing, in particular:

1. Anywhere that needs to log a diagnostic needs to be passed a `diag.Sink`, preferably as an argument and not hanging off `Plan` like it is today
2. Anywhere that can log a diagnostic needs to return `(controlFlow, error)` so that callers can determine whether or not to proceed.

After this commit, we can say that the step generator returning a non-nil `error` is a bug in Pulumi. We can't quite realize the full benefit of this (bubbling all the way to the CLI top-level and presenting an error requesting a bug report) but we can tackle that in subsequent PRs.

Here's a before and after comparison of this PR applied to a check failure:

Before:

![screen shot 2018-08-29 at 3 02 05 pm](https://user-images.githubusercontent.com/1871912/44820152-8b02c180-aba4-11e8-842b-180d7ed9abae.png)

After:

![screen shot 2018-08-29 at 3 01 34 pm](https://user-images.githubusercontent.com/1871912/44820158-8f2edf00-aba4-11e8-9307-d969ec0134cd.png)

cc @lukehoban

joeduffy commented:
Nice cleanup. Since you are tidying things up generally, do you think we will end up in an end state where it simply says

```
error: pulumi-nodejs:dynamic:Resource ... property state is 42`
```

without the Diagnostics: global: global stuff (which I find just adds noise), and the terminating `error: an error occurred while advancing the preview`?

swgillespie commented:
To be clear, I expect that the above goal will take a few PRs to achieve - there's a nontrivial amount of plumbing that needs to get done before we're in a place where we can get away with using the diagnostic sink to determine our exit status.

joeduffy commented:
> I expect that the above goal will take a few PRs to achieve

:+1: totally get it. Please just allow me to register enthusiasm for the expected end state :smile: 🎉

CyrusNajmabadi commented on pkg/resource/deploy/plan_executor.go:
```diff
@@ -36,9 +36,6 @@ type planExecutor struct {
 }
 
-// Utility for convenient logging.
-var log = logging.V(4)
```
how come?

swgillespie replied:
Turns out this doesn't work anymore - `logging.V(4)` gets the logging level at init, and we don't use the normal `glog` initialization mechanism (i.e. flags) anymore and set the log level after module constructors have run.

CyrusNajmabadi commented on pkg/resource/deploy/plan_executor.go:
```diff
@@ -148,19 +145,35 @@ func (pe *planExecutor) Execute(callerCtx context.Context, opts Options, preview
+					}
+
+					logging.V(4).Infof("planExecutor.Execute(...): canceling context")
 					cancel()
```
this was an issue before.  However, i think it would probably always be good to explain the implications of these sorts of calls as htey are essentially non local flow control as well.

swgillespie replied:
will do.

CyrusNajmabadi commented on pkg/resource/deploy/plan_executor.go:
```diff
@@ -184,29 +197,35 @@ func (pe *planExecutor) Execute(callerCtx context.Context, opts Options, preview
+	logging.V(4).Infof("planExecutor.handleSingleEvent(...): generated steps, err=%s, status=%d", err, flow)
 	if err != nil {
-		return err
+		return flow, err
```
this is both interesting and hard to reason about.  We got an error from a deeper path, but we're also passing along flow control dcisions from a deeper level to a higher one.  it's not clear to me that that's necessarily the right thing to do.

swgillespie commented:
Based on @CyrusNajmabadi's in-person and online feedback I've reworked the programming idioms here a bit.

`controlFlow` now is an enum with three variants:
* `Bail` - Indicates to the caller that the callee issued a diagnostic and does not think that the plan should continue.
* `Continue` - Indicates to the caller that the callee thinks the plan should continue. The callee *may still have issued diagnostics*, but the plan can still proceed despite it.
* `Undefined` - Indicates to the caller that the callee doesn't know what to do because it or something it calls transitively has hit an internal-to-Pulumi error (i.e. a bug). It's up to the caller to decide what to do.

A function returning a `controlFlow` returns two or three things:

1. First, if the function returns data (i.e. `[]Step`, as is the case with this PR), the first return value is the data.
2. The recommended control flow, one of `Bail`, `Continue`, or `Undefined`.
3. An `error`.

The returned data is not the default value, `controlFlow` must be `Continue` and `error` must be nil. Callers interpret this as "please continue, this operation was successful."

If `controlFlow` is `Bail`, the data must be the default value and `error` must be nil. Callers interpret this as "please do not continue, this operation was not successful but we have gracefully issued appropriate diagnostics."

If `controlFlow` is `Undefined`, the data must be the default value and `error` must not be nil. Callers interpret this as "figure out what to do next, this operation failed because of an internal error."

CyrusNajmabadi commented on pkg/resource/deploy/plan_executor.go:
```diff
@@ -148,19 +145,35 @@ func (pe *planExecutor) Execute(callerCtx context.Context, opts Options, preview
+					// error that ultimately bubbles up to the top level.
+					if eventErr == nil {
+						eventErr = errors.New("step generation failed")
+					}
```
this definitely seems like a reasonable way to 'snip' off the virality of this change at certain points.  However, in general, it would be nice if there was some type/pattern we could use here so you can go back later and move these forward.  i.e. have some sort of helper function (which simply makes an error) but now can be documetned and can be something we look for and remove later.

swgillespie replied:
Something like `context.TODO()` is for contexts? That would make sense here for sure

CyrusNajmabadi commented on pkg/resource/deploy/step_generator.go:
```diff
@@ -25,6 +25,40 @@ import (
+// used in tandem with an `error` return value. A function that returns a (controlFlow, error) means that
+// the called function can produce diagnostics and *gracefully fail*, while also being capable of failing ungracefully.
+// It is a *bug in Pulumi* whenever a function of this type returns a non-nil error, and it is expected that
+// callers of these functions propegate the returned errors upwards so that higher-level components can print
```
propegate

CyrusNajmabadi commented on pkg/resource/deploy/step_generator.go:
```diff
@@ -296,9 +330,9 @@ func (sg *stepGenerator) GenerateSteps(event RegisterResourceEvent) ([]Step, err
+						return nil, Undefined, err
 					} else if sg.issueCheckErrors(new, urn, failures) {
-						return nil, errors.New("One or more resource validation errors occurred; refusing to proceed")
+						return nil, Bail, nil
```
expected change?

swgillespie replied:
Yeah - I missed this one earlier. Sorry, I thought this made it into the first commit.

CyrusNajmabadi commented on pkg/resource/deploy/step_generator.go:
```diff
@@ -204,7 +238,7 @@ func (sg *stepGenerator) GenerateSteps(event RegisterResourceEvent) ([]Step, err
 	// If the resource isn't valid, don't proceed any further.
 	if invalid {
-		return nil, errors.New("One or more resource validation errors occurred; refusing to proceed")
+		return nil, Bail, nil
```
expected?

swgillespie commented:
The most recent commit introduces a `Result` type to encapsulate the error and control flow types in the original PR. The contents of this PR came from a couple brainstorms with Pat and Cyrus, so I'd appreciate eyes on the design if either of you have the time today.

The original before/after screenshots in this PR still apply here, so the end result of this PR is still the same.

swgillespie commented:
One thing that came up in our brainstorming is that it would be useful if `Error` could be called on a nil receiver, so that callers don't ever need to check `res != nil` before turning it into an error with `res.Error`. Cyrus suggested we move `Result` to an interface so that we're not passing pointers around everywhere, which I did in the most recent commit, but we lose the ability to invoke methods on null Results.

CyrusNajmabadi commented on pkg/util/result/result.go:
```diff
@@ -0,0 +1,85 @@
+// adapted to use Results.  Their use is intended to be temporary until Results
+// are plumbed throughout the Pulumi codebase.
+func TODO() error {
+	return errors.New("bailng due to error")
```
spelling.

CyrusNajmabadi replied:
can you add the comment you just explained to me about how this will *not* be in the users face all the time in the interim period?

swgillespie commented on pkg/resource/deploy/plan_executor.go:
```diff
@@ -153,11 +154,14 @@ func (pe *planExecutor) Execute(callerCtx context.Context, opts Options, preview
+					}
 					cancel()
-					return false, eventErr
+					return false, result.TODO()
```
Not sure if this is reasonable API design, but part of me wants to add a string (literal) argument to `result.TODO` that forces users to write a comment as to why this TODO is necessary. It could be an actual code comment here, todo.

pgavlin commented on pkg/resource/deploy/plan_executor.go:
```diff
@@ -153,11 +154,14 @@ func (pe *planExecutor) Execute(callerCtx context.Context, opts Options, preview
-					pe.reportError(pe.plan.generateEventURN(event.Event), eventErr)
+				if res := pe.handleSingleEvent(event.Event); res != nil {
+					resErr := res.Error()
+					if resErr != nil {
```
`if resErr := res.Error(); resErr != nil {`

pgavlin commented on pkg/resource/deploy/step_generator.go:
```diff
@@ -186,14 +185,14 @@ func (sg *stepGenerator) GenerateSteps(event RegisterResourceEvent) ([]Step, err
+			return nil, result.FromError(err)
 		} else if analyzer == nil {
-			return nil, errors.Errorf("analyzer '%v' could not be loaded from your $PATH", a)
+			return nil, result.Errorf("analyzer '%v' could not be loaded from your $PATH", a)
```
Should this instead report a diagnostic and then `Bail`?

pgavlin commented on pkg/util/result/result.go:
```diff
@@ -0,0 +1,85 @@
+// At the highest level, when a function wishes to return only an `error`, the
+// `Error` member function can be used to turn a nullable `Result` into an
+// `error`.
+type Result interface {
```
If `resultImpl` is the only expected implementor of this interface, then this is very non-idiomatic. Instead, we should just use a struct with methods on its pointer type:
```
type Result struct {
    err error
}

func (r *Result) Error() error {
    return r.err
}
```

pgavlin commented on pkg/util/result/result.go:
```diff
@@ -0,0 +1,85 @@
+
+// Bail produces a Result that represents a computation that failed to complete
+// successfully but is not a bug in Pulumi.
+func Bail() Result {
```
I don't love the term "bail", but I don't have a better suggestion offhand. Perhaps `Terminate` or `Done` or something similar? I could certainly live with "bail" if nothing better arises.

swgillespie replied:
I can't think of anything better than `Bail` - `Terminate` and `Done` are both already used by contexts so I think it would be confusing to overload those terms again in the codebase.

pgavlin replied:
`Abort`?

pgavlin replied:
Or even `Stop`?

pgavlin replied:
(I know that this is bikeshedding, but given that this is sort of an odd pattern I do want it to be as obvious and readable as possible)

swgillespie commented:
@pgavlin Addressed your feedback - I think this is ready to merge if there aren't any objections. I'm going to start pushing `result.Result` through more parts of the engine once this PR merges.

swgillespie commented:
I'm going to go ahead and merge this so I can get started on follow-up work. Let me know if there are any changes I should make.
